abstract_domains/numerical/octagon.py

```diff
@@ -0,0 +1,578 @@
+
+from core.expressions_tools import ExpressionVisitor, ExpressionTransformer, \
+    make_condition_not_free, simplify
+
+Sign = UnaryArithmeticOperation.Operator
+PLUS = Sign.Add
+MINUS = Sign.Sub
+
+
+def _index_shift(sign: Sign):
+    return 1 if sign == MINUS else 0
+
+
+class OctagonLattice(BottomMixin, NumericalMixin):
+    def __init__(self, variables: List[VariableIdentifier]):
+        """Create an Octagon Lattice for given variables.
+        
+        :param variables: list of program variables
+        """
+        self._variables = variables
+        self._var_to_index = {}
+        self._index_to_var = {}
+        index = 0
+        for var in self.variables:
+            self._var_to_index[var] = index
+            self._index_to_var[index] = var
+            self._index_to_var[index + 1] = var
+            index += 2
+        self._dbm = IntegerCDBM(len(variables) * 2)
+        super().__init__()
+
+        for key in self.dbm.keys():
+            self.dbm[key] = inf
+
+    @property
+    def variables(self):
+        return self._variables
+
+    @property
+    def dbm(self):
+        return self._dbm
+
+    def __getitem__(self, index_tuple: Tuple[Sign, VariableIdentifier, Sign, VariableIdentifier]):
+        if len(index_tuple) == 4:
+            sign1, var1, sign2, var2 = index_tuple
+            return self.dbm[
+                self._var_to_index[var1] + _index_shift(sign1), self._var_to_index[var2] + _index_shift(sign2)
+            ]
+        else:
+            raise ValueError("Index into octagon has invalid format.")
+
+    def __setitem__(self, index_tuple: Tuple[Sign, VariableIdentifier, Sign, VariableIdentifier], value):
+        if len(index_tuple) == 4:
+            sign1, var1, sign2, var2 = index_tuple
+            i, j = self._var_to_index[var1] + _index_shift(sign1), self._var_to_index[var2] + _index_shift(sign2)
+            if i != j:
+                self.dbm[i, j] = value
+        else:
+            raise ValueError("Index into octagon has invalid format.")
+
+    def binary_constraints_indices(self, sign1: Sign = None, var1: VariableIdentifier = None,
+                                   sign2: Sign = None, var2: VariableIdentifier = None):
+        signs1 = [sign1] if sign1 else [PLUS, MINUS]
+        signs2 = [sign2] if sign2 else [PLUS, MINUS]
+        vars1 = [var1] if var1 else self.variables
+        vars2 = [var2] if var2 else self.variables
+        for var1 in vars1:
+            for var2 in vars2:
+                if var1 == var2:
+                    # do not yield diagonal and upper right triangular matrix
+                    continue
+                for sign1 in signs1:
+                    for sign2 in signs2:
+                        yield (sign1, var1, sign2, var2)
+
+    def __repr__(self):
+        if self.is_bottom():
+            return "BOTTOM"
+        elif self.is_top():
+            return "TOP"
```

Comment:
Why don't we use the unicode characters `U+22A5` (bottom) and `U+22A4` (top) instead?